### DISCUSSION TITLE
✨ Add extension to deploy a container from a registry to a Kubernetes cluster

### DISCUSSION THREAD

**Discussion by vikram-dagger** [dagger/dagger#12078] [Ideas]
### What are you trying to do?

Assume that I have built an application image with Dagger/Cloak and pushed this to a registry like Docker Hub. As part of my deployment process, I would like to deploy this image from the registry to a K8s cluster. This could be either a local minikube cluster for testing, or a cloud K8s provider like GKE.

### Why is this important to you?

Being able to deploy to K8s should be supported in Dagger as many deployments now happen on K8s.

**Comment by jpadams** [comment]
@bluebrown could you advise us on what this could look like from Dagger/Cloak to AKS?

Guessing we can start with this example your submitted using `kubectl`:
https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/alpha/azure/test/test.cue

or this one with `helm`:
https://github.com/dagger/dagger/blob/main/pkg/universe.dagger.io/alpha/azure/aks/helm/test/test.cue

WDYT?

**Comment by bluebrown** [comment]
@vikram-dagger, Kubernetes pulls images by itself. So you are not giving it the image really via dagger. You would only tell it where it can get the image.

The most simple thing you could do would be mounting your local kubeconfig to a kubectl image and apply some manifest referencing your image in some pods container list.

Once you have an action in place to run a command against your cluster, you need to ensure that your cluster is allowed to pull from the registry in case of private registry/repo. One way of doing this is using image pull secrets: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/.

At the moment, there is no simple kubectl command in the universe. I wanted to make a PR for something like the below action, but I did not get around to do it, yet. However, you can use this as starting point or example:

```c
package main

import (
	"dagger.io/dagger"
	"dagger.io/dagger/core"
	"universe.dagger.io/docker"
)

dagger.#Plan & {
	client: {
		env: KUBECONFIG: *"~/.kube/config" | string
		commands: kubeconfig: {
			name: "cat"
			args: ["\(env.KUBECONFIG)"]
			stdout: dagger.#Secret
		}
	}
	actions: {
		project: core.#Source & {
			path: "."
		}
		deploy: docker.#Run & {
			always: true
			_img:   docker.#Pull & {
				source: "bitnami/kubectl"
			}
			input: _img.output
			command: {
				name: "apply"
				args: ["--namespace", "sandbox", "--filename", "kube/manifests/"]
			}
			user:    "root"
			workdir: "/workspace"
			mounts: {
				"/.kube/config": {
					dest:     "/.kube/config"
					type:     "secret"
					contents: client.commands.kubeconfig.stdout
				}
				"/workspace": {
					contents: project.output
					dest:     "/workspace"
				}
			}
		}
	}
}

```
